### ISSUE TITLE
Flatpak version - firefoxpwa

### ISSUE THREAD

**Issue by realgrm** [filips123/PWAsForFirefox#142]
### Proposed Solution

- Plugin. [Example](https://github.com/flathub/com.obsproject.Studio.Plugin.OBSVkCapture). 
- flatpak-spawn. [Example](https://github.com/flathub/com.ulduzsoft.Birdtray/blob/master/com.ulduzsoft.Birdtray.json). 
- Maybe more solutions arise asking questions for the developers, check below

### Additional Information

You can ask questions to the flatpak/Flathub developers by:
- [Chat - Matrix](https://matrix.to/#/#flatpak:matrix.org)
- [Forum - discourse.flathub](https://discourse.flathub.org/)

To search for more examples:
- [Plugins available on Flathub](https://github.com/search?q=org%3Aflathub+Plugin&type=repositories)
- [flatpak-spawn](https://github.com/search?q=org%3Aflathub+flatpak-spawn&type=code)

**Comment by realgrm** [comment]
In case of discoverability of the app in case of being implemented as plugin (add-on) in the Gnome Software:
Example: com.obsproject.Studio.Plugin.OBSVkCapture

- search results
<img src="https://user-images.githubusercontent.com/23300290/161197837-99ab47fd-8c44-4561-9a6c-ceb6107fb2ad.png" width="50%" >

- installation
<img src="https://user-images.githubusercontent.com/23300290/161197106-3a109442-d675-42cb-a1e8-112fc86ca373.png" width="50%" >

**Comment by filips123** [comment]
Thank you for providing detailed suggestions on how to implement Flatpak support, but I still have few questions.

Will distributing PWAsForFirefox native app as a plugin allow Firefox to access its [Native Messaging manifests](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests)? The problem is that they currently need to be present in a specific directory so Firefox can found them. However, there seems to be some ongoing work to support Native Messaging in sandboxed browsers (flatpak/xdg-desktop-portal#655, flatpak/xdg-desktop-portal#705) which will probably help.

Another problem is that the Firefox runtime needs to be modified a bit with [some additional files](https://github.com/filips123/PWAsForFirefox/tree/main/native/userchrome/runtime) so it can load all modifications. Is it possible to do this as a Flatpak plugin without affecting the main Firefox? Maybe using FUSE filesystem (#67) with Flatpak's Firefox as a `lowerdir` and some PWAsForFirefox-specific directory as a merged dir? Currently, PWAsForFirefox just downloads its own Firefox runtime from Mozilla and modifies it. However, this means Firefox is duplicated on the system (one system Firefox and one PWA Firefox), and it's probably not a best practice in the Flatpak world.

**Comment by VortexAcherontic** [comment]
Hello everyone, sorry for bumping this thread without supplying any valuable information but I have a few questions regarding this project.

Had some hopes to get PWA support on Firefox Flatpak working as well and was questioning why is the setup so much different compared to chrome based browsers?  
I am just a novice regarding the entire topic and not an expert in any browser so I apologize for my ignorance.  
On chrome based browsers (Edge, Chromium, Chrome and even Gnome Web) some can just save the web page as an app and it instantly shows up on your desktop or app launcher also from inside a flatpak sandbox.  

Or is it that those browser just spawn a minimal browser window without address bar and significantly less buttons and menus and call it an "app" while this project and Firefox aim to do it right?

Additionally I run an immutable Linux distribution (openSUSE MicroOS) which makes installing packages into the host not impossible but requires extra steps and a reboot.  
With the rise of the Steam Deck and SteamOS 3 I see this to be a pressuring issue some time soon.  
Also on distributions sch as Fedora Silverblue and Onyx I foresee this to be an issue.  
I would like to see an implementation not requiring additional (unsigned) packages to be install in the host operating system.

Sure I could run Firefox entirely from a Distobox container and also provide the PWA support there but on the other hand this will partially break the desktop integration with the host OS for numerous reasons which are out of scope for this comment.

Anyway the question remains why the Fierfox way of doing this is so much different from the chrome way?

Many thanks for reading this and kind regards,  
V.

**Comment by filips123** [comment]
> why is the setup so much different compared to chrome based browsers?

Because Chrome has built-in PWA support while Firefox does not, so this project needs to be installed as a separate extension and additional native program. Just extension is not enough as web extensions do not have the ability to do all things that are needed for PWA support, so an additional native program is required.

> [...] while this project and Firefox aim to do it right?

I mean, Chrome has working support for PWAs and I can't say they haven't "done it right". However, this project also offers some features that AFAIK Chromium browsers do not:

* Support for separate profiles (although this is just a by-product of how this project is made). Sometimes it may be more convenient to use the same profile for PWAs and main browsing (as it's done on Chromium), but using separate profiles allows you to do more things that aren't possible on Chromium, like extra customization or using multiple accounts for the same PWA.

* More customization (partially related to the previous topic). This project supports additional configuration abilities that Chromium browsers do not. For example, allowing customization of the browser toolbar, different browser settings for different PWAs, additional addons, etc.

* Better integration with the desktop. Although Chrome and Edge support quite good integration with system (protocol handlers, shortcuts, etc.), last time I checked this was limited to Windows (while on Linux they only created a single desktop shortcut). This project also supports that additional integration on Linux.

> I would like to see an implementation not requiring additional (unsigned) packages to be install in the host operating system.

I would also like this, but currently it's impossible because Flatpak does not support the required features to make this work yet. It will probably change in the future because they're working on this on the Flatpak side.

Also, what do you mean by (un)signed packages? Do you mean just publishing this to Flathub or some additional signing certificates, like those on Windows and macOS?

> Anyway the question remains why the Fierfox way of doing this is so much different from the chrome way?

TLDR: This project is Firefox extension because Firefox does not have built-in PWAs and an additional native program is required to make it work. Although this causes some problems (like harder installation), it also allows some features that are not available on Chromium browsers (multiple profiles, better customization, better integration). Flatpak is currently not an option because it does not support all required features, but that will probably change in the future.

**Comment by VortexAcherontic** [comment]
Many thanks for the detail explanation. Now things are a lot more clear to me and I feel like I now properly understand the situation ❤️ 

> Also, what do you mean by (un)signed packages?

Sorry for the confusion here.  
I now see, after reading the instructions again, that this was may fault as I downloaded the RPM directly instead of adding the package cloud repo.  
Hence my package manager complained about an unsigned RPM to be installed. My bad.

Will keep an eye on the Flatpak side of things for now.

Thank you for your efforts as it does not seem like an easy undertaking getting all this stuff working.

Kind regards,
V.

**Comment by ar-lex** [comment]
I have some thoughts and questions after experimenting with native app and flatpak version of Firefox.

**flatpak-nspawn** method works. I just installed the system package (`.deb` in my case) and then

- created a `$HOME/.var/app/org.mozilla.firefox/data/firefoxpwa.sh` script with a single command `flatpak-spawn --host /usr/libexec/firefoxpwa-connector "$@"`
- copied `/usr/lib64/mozilla/native-messaging-host/firefoxpwa.json` to `$HOME/.var/app/org.mozilla.firefox/.mozilla/native-messaging-host/` and changed `/usr/libexec/firefoxpwa-connector` to full path of the created script.

That's it, everything works as expected - Firefox in flatpak spawns the unconfined binary which generates runtime, profile, `.desktop` files and icons, then starts that modified Firefox also unconfined.

Then I tried to run the native app confined:

- extracted the contents of a `.deb` package to `$HOME/.var/app/org.mozilla.firefox`, specifically
  - `/usr/bin/firefoxpwa` and `/usr/libexec/firefoxpwa-connector` to `$HOME/.var/app/org.mozilla.firefox/data/bin/`
  - `/usr/share/firefoxpwa/userchrome` to `$HOME/.var/app/org.mozilla.firefox/data/share/`
  - `firefoxpwa.json` similarly to `flatpak-nspawn` method, setting the full path for extracted `firefoxpwa-connector`
- modified flatpak permissions (I used Flatseal):
  - added filesystem access for `xdg-data/applications` and `xdg-data/icons` so that the native app could create `.desktop` files and icons (which need to be additionally modified, but it's easier anyway)
  - set `FFPWA_SYSDATA` and `FFPWA_EXECUTABLES` to full paths accordingly

With these modification the browser extension detects it's native part. The runtime is succesfully downloaded and modified, the app profiles and `.desktop` files are generated, but the app can't be correctly launched. I used the following command to start the app:
```
flatpak run --command=/home/<username>/.var/app/org.mozilla.firefox/data/bin/firefoxpwa org.mozilla.firefox site launch <AppID> --protocol
```
No errors in firefoxpwa logs and no stdout/stderr messages, I only get the following error dialog window:

> Failed to read the configuration file. Please contact your system administrator.

As I understood, the `_autoconfig.cfg` isn't loaded successfully - when I remove `general.config.filename` from `autoconfig.js`, there is no such error, but obviously no modifications either. Any assistance welcome.

Regarding other suggestions, particularly flatpak plugin, I don't think that it is the optimal solution. The browser itself has to be modified - Firefox flatpak packaging manifest has to have the definitions for extension file paths and other conventions. This extra complexity IMHO doesn't give any security advantages.

I think there is no need to wait for native messaging portal implementation in flatpak either. Everything we need should be already here. Instead we should aim for the simplicity of [VideoDownloadHelper extension approach](https://github.com/aclap-dev/vdhcoapp/issues/137#issuecomment-1544985068), which I used as an example for my experiments. If the problem that I encountered would be solved, then the confined native app installation could be as easy as extracting the archive and running the script, everything as a non-privileged user.

**Comment by filips123** [comment]
Maybe RPM version on Fedora is compiled with different compiler optimizations?

**Comment by maymage** [comment]
Possibly related

* https://github.com/flatpak/xdg-native-messaging-proxy

**Comment by filips123** [comment]
That native messaging proxy will allow Firefox packaged in Flatpak or Snap to communicate with PWAsForFirefox installed natively. However, it still won't allow PWAsForFirefox to be packaged in Flatpak. Support for that is planned in Flatpak but won't be "available anytime soon". More info [here](https://github.com/flatpak/xdg-desktop-portal/issues/655#issuecomment-3130052988) and in related comments.

**Comment by NeatNit** [comment]
> That native messaging proxy will allow Firefox packaged in Flatpak or Snap to communicate with PWAsForFirefox installed natively. However, it still won't allow PWAsForFirefox to be packaged in Flatpak.

As things stand *today*, can one use Flatpak firefox with non-Flatpak PWAsForFirefox?

Sorry if this has been answered before, I haven't seen a clear answer.

**Comment by filips123** [comment]
> As things stand _today_, can one use Flatpak firefox with non-Flatpak PWAsForFirefox?

Appearently, `xdg-native-messaging-proxy` is now available [in some distros](https://repology.org/project/xdg-native-messaging-proxy/versions), but you probably need to manually install it.

However, it still hasn't been implemented in Firefox (BZ-1955255), so unfortunately you can't use it yet until that is implemented.

**Comment by dnkmmr69420** [comment]
Do you plan to implement it soon? it would be nice especially for Bazzite users and users of other forms of Fedora Atomic.

**Comment by NeatNit** [comment]
> Do you plan to implement it soon?

From what I can tell as an external observer, it's out of Filip's hands. Flatpak doesn't yet support the features needed to make it work. I'm sure PWAsForFirefox will be the first to make use of it when it finally becomes possible.

**Comment by filips123** [comment]
Yes, NeatNit is correct. Unfortunately, there's nothing I can do until support for that is implemented in Flatpak and Firefox.

For using natively-installed PWAsForFirefox from Flatpak-installed Firefox, there is some progress going on (BZ-1955255, PHAB-242328), so that should hopefully be available soon. However, keep in mind that once that patch lands in Firefox, there will be around another two months before it's available in a stable release. So, soon probably means this year.

For using Flatpak-installed PWAsForFirefox, don't expect anything at least for the foreseeable future. The progress to support that hasn't even been started...
